### PR TITLE
Mk v1.1

### PR DESCRIPTION
**Pull request by mkao006**

### CODE REVIEW

File: pipeline/price_modelling/controller.py [hunk 1]
```diff
@@ -1,147 +1,23 @@
...
 import os
-import pandas as pd
-import numpy as np
 from sqlalchemy import create_engine
-from datetime import timedelta
-from sklearn.linear_model import ElasticNetCV
-from statsmodels.nonparametric.smoothers_lowess import lowess
+import pandas as pd
+
+forecast_period = 90
```
Comment (marcogarieri) [inline pipeline/price_modelling/controller.py]: Probably you already told me, but have you already tested different lags?
Should we do, after the meeting, a proper test to tune this?

File: pipeline/price_modelling/processor.py [hunk 2]
```diff
@@ -1,42 +1,19 @@
...
-smoothed_prediction = (
-    ctr.train_bag_elasticnet(complete_data=transformed_data,
-                             forecast_period=forecast_period,
-                             holdout_period=holdout_period,
-                             bootstrapIteration=bootstrapIteration,
-                             topic_variables=topic_variables,
-                             response_variable=response_variable))
+elasticnet_prediction = mbe.output()
+lstm_prediction = mlks.output()
+all_prediction = pd.concat([elasticnet_prediction, lstm_prediction])
```
Comment (marcogarieri) [inline pipeline/price_modelling/processor.py]: Always clean code! Neat job!!

File: pipeline/price_modelling/modeler_bagged_elasticnet.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    train_data = model_data.iloc[1:-(holdout_period - 1)]
+    test_data = model_data.iloc[-(holdout_period):]
+    cutoff_date = train_data['date'].max() + timedelta(days=forecast_period)
+    return prediction_data, train_data, test_data, cutoff_date
+
+
+def train_bag_elasticnet(complete_data, forecast_period, holdout_period,
+                         bootstrapIteration, topic_variables,
+                         response_variable, date_col='date'):
+    '''Function to bag and train the Elastic net iwth cross-validation
```
Comment (marcogarieri) [inline pipeline/price_modelling/modeler_bagged_elasticnet.py]: just a small type "iwht"

File: pipeline/price_modelling/modeler_bagged_elasticnet.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        build_datasets(complete_data=complete_data,
+                       forecast_period=forecast_period,
+                       holdout_period=holdout_period,
+                       response_variable=response_variable))
+
+    observation_length = complete_data.shape[0]
+    total_variable_count = len(topic_variables)
+    predictions = np.zeros(shape=(observation_length, bootstrapIteration))
+    cv_min = np.zeros(bootstrapIteration)
+    # This is the of the implementation in R due to the specification of Python
```
Comment (marcogarieri) [inline pipeline/price_modelling/modeler_bagged_elasticnet.py]: ??

File: pipeline/price_modelling/modeler_lstm_k_step.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+from datetime import timedelta
+from sklearn import preprocessing
+import matplotlib.pyplot as plt
+from statsmodels.nonparametric.smoothers_lowess import lowess
+import controller as ctr
+
+topicModelTable = 'TopicModel'
+model_start_date = datetime(2010, 1, 1).date()
+
+# Model parameters
```
Comment (marcogarieri) [inline pipeline/price_modelling/modeler_lstm_k_step.py]: I remember you mentioned that you were working in best selection of hyper parameters. You used any Bayesian optimiser to find the best?

File: pipeline/price_modelling/modeler_lstm_k_step.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+
+
+def get_topic_modelled_article():
+    data_dir = os.environ['DATA_DIR']
+    engine = create_engine(
+        'sqlite:///{0}/the_reading_machine.db'.format(data_dir))
+    topic_modelled_article = pd.read_sql(
+        'SELECT * FROM {}'.format(topicModelTable), engine)
+    sclae_input = topic_modelled_article.drop('id', axis=1)
+    scaled_topic = pd.DataFrame(preprocessing.scale(sclae_input),
```
Comment (marcogarieri) [inline pipeline/price_modelling/modeler_lstm_k_step.py]: That's a question more than a comment. Wouldn't be better to use `preprocessing.StandardScaler()`, to apply the same transformation of the training set to the test set?

**Comment by mrpozzi** [comment]
LGTM

File: pipeline/price_modelling/modeler_bagged_elasticnet.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+def get_topic_variables():
+    ''' Function to get all the names of the topic model.
+    '''
+
+    topic_variables = (
+        pd.read_sql('PRAGMA table_info(TopicModel)', engine)['name']
+        .where(lambda x: x != 'id')
+        .dropna()
+        .tolist())
+    cmoplete_topic_variable = [v + suffix
```
Comment (mrpozzi) [inline pipeline/price_modelling/modeler_bagged_elasticnet.py]: nit. typo

File: pipeline/price_modelling/modeler_bagged_elasticnet.py (hunk 4 above)
Comment (mrpozzi) [reply pipeline/price_modelling/modeler_bagged_elasticnet.py]: ???

File: pipeline/price_modelling/modeler_lstm_k_step.py [hunk 8]
```diff
@@ -0,0 +1,505 @@
+import os
```
Comment (mrpozzi) [inline pipeline/price_modelling/modeler_lstm_k_step.py]: This file is huge.. any chance it could be split in subroutines?

File: pipeline/price_modelling/controller.py (hunk 1 above)
Comment (mkao006) [reply pipeline/price_modelling/controller.py]: I have, but it's inconclusive. I guess the question is how far into the future with what level of uncertainty is a good mix for the purpose of the prediction.

But this is something that we do need to consider and adjust later.

File: pipeline/price_modelling/modeler_bagged_elasticnet.py (hunk 4 above)
Comment (mkao006) [reply pipeline/price_modelling/modeler_bagged_elasticnet.py]: I guess @marcogarieri 's question was about the comment, but at least elaborate it....

My comment was poorly phrased and hastily written. Basically, the parameterisation of the exponential distribution is different in R as opposed to Python. By default, R uses `rate` while Python uses `scale`. Thus the reciprocal value of R is taken for Python so that the two implementation match.

File: pipeline/price_modelling/modeler_lstm_k_step.py (hunk 5 above)
Comment (mkao006) [reply pipeline/price_modelling/modeler_lstm_k_step.py]: Not for the Elasticnet, the model doesn't really have hyperparameters. The `filter_coef` is the amount of exponential smoothing, but after several trial and error, the value 1 performs significantly better (actually it is a `cumsum`). For the `bootstrapIteration`, you generally don't need to have a large sample to reduce the variance.

The selection of hyper-parameter is much more relevant for the RNN, but currently we run through a grid of values and examine the result on Tensorboard.

There are methods which are designed to specifically optimise the hyperparameters of an RNN. Please see the [learning to learn](https://arxiv.org/pdf/1606.04474.pdf), but Bayesian optimiser is also another popular option we can explore.

P.S. This section actual outlines one of the main difference between the ML/DL approach. For Elasticnet, I had to find some transformation in the data so the model predicts well, on the other hand, I didn't have to transform the data (other than normalisation) but the tuning of the model requires significant work.

File: pipeline/price_modelling/modeler_lstm_k_step.py (hunk 6 above)
Comment (mkao006) [reply pipeline/price_modelling/modeler_lstm_k_step.py]: Not sure what you meant here, but the transformation is applied to the input set before the data is split into training and test set. So the same transformation is applied, also, they need to be scaled together, since we are working with time-series where the scale should be the same across time.

File: pipeline/price_modelling/modeler_bagged_elasticnet.py (hunk 7 above)
Comment (mkao006) [reply pipeline/price_modelling/modeler_bagged_elasticnet.py]: corrected.

File: pipeline/price_modelling/modeler_lstm_k_step.py (hunk 8 above)
Comment (mkao006) [reply pipeline/price_modelling/modeler_lstm_k_step.py]: The first 120 lines are data processing which should have been in a separate data processing module. 

I have not done so for two reasons, 1. I was in a hurry to get the model working, so I didn't want to work on 2 separate module. 2. The whole data processing module may change since phase two requires a complete different pipeline, and I didn't have a good idea how I want to restructure it.

Eventually, we will definitely restructure the input pipeline for the RNN and remove the chunk.

Comment (mrpozzi) [reply pipeline/price_modelling/modeler_lstm_k_step.py]: I figured it was due to practical concerns, but maybe we can track it with a TODO..

File: pipeline/price_modelling/modeler_bagged_elasticnet.py (hunk 4 above)
Comment (mrpozzi) [reply pipeline/price_modelling/modeler_bagged_elasticnet.py]: yeah, we were both perplexed about the comment.
Makes total sense, maybe copy what you just wrote here in the comment?
